content/updates/terraform-null-label.md

```diff
@@ -0,0 +1,194 @@
+
+Enter `terraform-null-label`.
+
+Here’s an example infrastructure, with `terraform-null-label` applied:
+
+**NOTE:** This is a simplistic take on this configuration. Real production ready configurations are often _much_ more complex, with many more resources at play, but you can already see the potential for reuse and the time and consistency wins with `terraform-null-label`.
+
+```hcl
+  module "public_alb_label" {
+    source  = "cloudposse/label/null"
+    version = "0.25.0"
+
+    namespace   = "mp"
+    stage       = "prod"
+    environment = "uw2"
+    name        = "alb"
+    attributes  = ["public"]
+
+    tags = {
+      BusinessUnit       = "ENG"
+      Team               = "Reporting"
+      ManagedByTerraform = "True"
+    }
+  }
+
+  # ALB
+  resource "aws_lb" "default" {
+
+    # ID is the concatenation result of all ID element inputs i.e. `mp-uw2-prod-alb-public`
+    name               = module.public_alb_label.id
+
+    internal           = false
+    load_balancer_type = "application"
+    security_groups    = [aws_security_group.alb_sg.id]
+    subnets            = ["subnet-abc12345", "subnet-xyz67890"]
+
+    # terraform-null-label outputs the merged tags input and all ID element inputs as a map value
+    # i.e. { namespace: "mp", stage: "prod", Team: "Reporting", ManagedByTerraform: "True", ... }
+    tags               = module.public_alb_label.tags
+  }
+
+  resource "aws_security_group" "alb_sg" {
+    name        = module.public_alb_label.id
+    description = "Allow inbound traffic over HTTPS"
+    vpc_id      = "vpc-12345678"
+    tags        = module.public_alb_label.tags
+
+    ingress {
+      from_port   = 443
+      to_port     = 443
+      protocol    = "tcp"
+      cidr_blocks = ["0.0.0.0/0"]
+    }
+  }
+```
+
+It’s worth noting that in our examples, we’re using the AWS provider, but the label module can be used anywhere in Terraform regardless of provider or use-case.
+
+In the above example, the `terraform-null-label` module lets engineers define some common-sense values for the ID elements like stage and namespace, and then it can be used to apply a consistent name and tagging scheme to every resource and/or module that the team uses. "Write once, read many" and DRY (Don’t repeat yourself) are important factors we talk a lot about within good application code, and null-label enables us to easily incorporate those ideas into our platform code.
+
+Now, obviously that’s a very simplified and abstract example. In complex configurations, there are likely going to be many resources and modules with different usage, and in this case even with `terraform-null-label` it can start to get cumbersome to have to define new label modules each time you need to have variations in your ID elements. There is a solution to that however as one of the module’s most powerful features addresses exactly that: the `context` input + output value.
```

Comment:
It may be worth mentioning that this is just taking a common behavior from languages like Perl or JavaScript, that typically pass around something called a "context." 

Just for clarification, and this doesn't need to be included in the article: In older JavaScript patterns, particularly before the ES6 arrow functions, the this context is sometimes explicitly passed as the first argument using methods JavaScript. In Perl, context is often conveyed through a `self` variable passed to methods. The context provides information that helps a function or method operate in a particular scope or environment.

e.g. 
```package Person;

sub new {
    my $class = shift;
    my $self = {
        name => shift,
    };
    bless $self, $class;
    return $self;
}

sub say_hello {
    my ($self) = @_;
    print "Hello, my name is $self->{name}\n";
}

# Main program
my $person = Person->new("John");
$person->say_hello();  # Output: "Hello, my name is John"
```


e.g. in JavaScript
```
function sayHello(greeting) {
    console.log(`${greeting}, my name is ${this.name}`);
}

const obj = {
    name: "John"
};

sayHello.call(obj, "Hello");  // Output: "Hello, my name is John"
```

Reply:
Ah great callout @osterman, I've seen the same pattern in other languages as well, but I never made the connection to null-label context. 

We found this post was getting a bit long, so we're going to follow up with a whole post on `context.tf` (and maybe mixins in general). When we do, we'll be sure to include this as I'm sure it'll help people get what is going on here!